manual/src/tutorials/moduleexamples.etex

```diff
@@ -13,51 +13,43 @@ confusing names. Such a package is called a {\em structure} and
 is introduced by the "struct"\ldots"end" construct, which contains an
 arbitrary sequence of definitions. The structure is usually given a
 name with the "module" binding. For instance, here is a structure
-packaging together a type of priority queues and their operations:
+packaging together a type of FIFO queues and their operations:
 \begin{caml_example}{toplevel}
-module PrioQueue =
+module Fifo =
   struct
-    type priority = int
-    type 'a queue = Empty | Node of priority * 'a * 'a queue * 'a queue
-    let empty = Empty
-    let rec insert queue prio elt =
-      match queue with
-        Empty -> Node(prio, elt, Empty, Empty)
-      | Node(p, e, left, right) ->
-          if prio <= p
-          then Node(prio, elt, insert right p e, left)
-          else Node(p, e, insert right prio elt, left)
-    exception Queue_is_empty
-    let rec remove_top = function
-        Empty -> raise Queue_is_empty
-      | Node(prio, elt, left, Empty) -> left
-      | Node(prio, elt, Empty, right) -> right
-      | Node(prio, elt, (Node(lprio, lelt, _, _) as left),
-                        (Node(rprio, relt, _, _) as right)) ->
-          if lprio <= rprio
-          then Node(lprio, lelt, remove_top left, right)
-          else Node(rprio, relt, left, remove_top right)
-    let extract = function
-        Empty -> raise Queue_is_empty
-      | Node(prio, elt, _, _) as queue -> (prio, elt, remove_top queue)
+    type 'a queue = { front: 'a list; rear: 'a list }
+    let make_queue front rear =
+      match front with
+      | [] -> { front = List.rev rear; rear = [] }
+      | _  -> { front; rear }
+    let empty = { front = []; rear = [] }
+    let is_empty = function { front = [] } -> true | _ -> false
+    let add x q = make_queue q.front (x :: q.rear)
+    exception Empty
+    let top = function
+      | { front = [] } -> raise Empty
+      | { front = x :: _ } -> x
+    let pop = function
+      | { front = [] } -> raise Empty
```

Comment:
Implementation looks correct to me, except for the dreaded warning 9 activating for the incomplete record pattern matches (on by default when cutting-and-pasting this into a dune project in development mode).

```
Error (warning 9 [missing-record-field-pattern]): the following labels are not bound in this record pattern:
rear
```

Would it be slightly clearer to just use a tuple for `'a queue` instead of a record here, since the purpose is to demonstrate the module system?

Reply:
For the record: I think it is stupid to enable warning 9 by default.  Now, since this ship has sailed a while ago, I just pushed a version that does not trigger warning 9.  I tried using a tuple, but then the inferred types don't mention `'a queue` and are full of `'a list * 'a list`, which is not nice.